### ISSUE TITLE
Problem with container:reorder ?

### ISSUE THREAD

**Issue by DaveInDev** [jtackaberry/rtk#15]
Hi,
(I'm binbinhfr on reaper forum)
Working with my list of files, I bumped into a problem with container:reorder.
I look at the code, and found something strange :
in line   576 you have this :    targetidx = targetidx - 1  (if moving down into the table)

I'm not sure that it should be there, because table.insert always inserts before the given position.
example here :
````
function move(t,n1,n2)
  print("step1 ", table.unpack(t))
  local obj = table.remove(t,n1)
  print("step2 ", table.unpack(t))
  table.insert(t,n2,obj)
  print("step3 ", table.unpack(t))
end

move({"a", "b", "c", "d", "e"}, 3, 4)  --> a       b       d       c       e
move({"a", "b", "c", "d", "e"}, 1, 5)  --> b       c       d       e       a
move({"a", "b", "c", "d", "e"}, 2, 1)  --> b       a       c       d       e
move({"a", "b", "c", "d", "e"}, 3, 6)  --> a       b       d       e   crash...
````

the first and second calls do the job without substracting 1 to n2.
and the last call crashes because you cannot give to table.insert an index > #t.
so your clamping to #self.children + 1 just after is also strange.

but maybe something is not clear to me...

the code that finally works for me in a RTK viewport list is : 
````
function rtk.Container:reorder(widget, targetidx)
  local srcidx = self:get_child_index(widget)
  if srcidx ~= nil and srcidx ~= targetidx then
      local widgetattrs = table.remove(self.children, srcidx)
      if targetidx > #self.children then
        table.insert(self.children, widgetattrs)
      else
        table.insert(self.children, rtk.clamp(targetidx, 1, #self.children), widgetattrs)
      end
      self._child_index_by_id = nil
      self:queue_reflow(rtk.Widget.REFLOW_FULL)
      return true
  else
      return false
  end
end
````

**Comment by jtackaberry** [comment]
The reason I'm clamping to `#self.children + 1` is because that is the index needed to append to the end of the table.  Your example isn't working because, indeed, after you call table.remove(), n2=6 exceeds the bounds allowed by `table.insert()`.  But my implementation of `rtk.Container:reorder()` evaluates `#self.children` *after* `table.remove()`.   Consequently, `#self.children+1` is a valid index for `table.insert()` at that point.

The implementation considers targetidx as the index in current list before the removal.  So if you have A, B, and C children and you want B to go after C, before the move that's at index 4, so you'd use 4 as targetidx.  The behavior here is technically working as designed, but I'll be honest, I can't recall why I implemented it this way (probably going back 4 years now for this particular logic -- I use this function in Reaticulate for drag-and-drop reordering of banks assigned to the track), and I definitely agree it doesn't make sense and it needs to be changed.

The implementation can be a bit simpler than what you have:

```lua
function rtk.Container:reorder(widget, targetidx)
    local srcidx = self:get_child_index(widget)
    if srcidx ~= nil and srcidx ~= targetidx then
        local widgetattrs = table.remove(self.children, srcidx)
        table.insert(self.children, rtk.clamp(targetidx, 1, #self.children + 1), widgetattrs)
        self._child_index_by_id = nil
        self:queue_reflow(rtk.Widget.REFLOW_FULL)
        return true
    else
        return false
    end
end
```

The challenge I face now is that this is a breaking API change.  Existing scripts that depended on these weird reorder() semantics will break if I fix the logic.

I think perhaps I could try my luck here.  There aren't very many scripts using rtk yet, and of the handful I skimmed (at least those I know about), none of them seem to be using reorder().  So I just might get away with fixing it without pissing anyone off. :)

**Comment by DaveInDev** [comment]
ok, for the clamping. I had to read some more doc about LUA to understand that LUA table size can be more than its actual content's size. Quite disturbing, but ok. ;-)
I also understood that you should have a good reason to implement it with this logic, because obviously your Rearticulate is working with it. But I had a hard time understanding why my use of reorder was not working at all when reordering to higher positions. ;-)
If it's not a problem for you, you can left it as it is. Thanks to LUA, I have redefined your function rtk.Container:reorder(widget, targetidx) in my script, so that it works now more accordingly with my logic :-)

**Comment by jtackaberry** [comment]
>  I had to read some more doc about LUA to understand that LUA table size can be more than its actual content's size. Quite disturbing, but ok. ;-)

Depends what you mean by this.  `#t` does reliably return the number of positional elements in table `t`.  The index taken by table.insert() supports exactly 1 greater than `#t`  for the append case.  I think that's all that's relevant here.

Lua tables -- somewhat uniquely as far as I can tell, compared to similar data structures in other languages -- simultaneously support both positional elements (like arrays/lists) *and* named elements (like maps/dictionaries).  `#t` only refers to the count of positional elements, not named elements, so this may be what you were reading about.  I wouldn't call it disturbing, but it was definitely a surprise for me when I was first learning Lua.  Now that I have more experience with Lua under my belt, I understand this design decision, and in fact rtk exploits it in places.

> If it's not a problem for you, you can left it as it is. 

No, I definitely agreed it was confusing and should be changed.  Maybe that got lost in translation?  I did commit the fix last night, which is why this issue got closed.  This change will land in rtk v1.3.0, which, unless I find some showstoppers in final testing, I will release today.

> Thanks to LUA, I have redefined your function rtk.Container:reorder(widget, targetidx) in my script, so that it works now more accordingly with my logic :-)

Just beware that your implementation of `reorder()` will break `reorder_before()` and `reorder_after()`.  You can see in the commit linked above that I did also update these functions to adapt to the change in `reorder()`'s logic.  But anyway, you won't need to monkeypatch rtk with this change once you switch to the upcoming v1.3.0.

**Comment by DaveInDev** [comment]
oh, I did not realize about this #t+1 used for adding at the end of the queue, because it does not seem to be described here : 
https://www.lua.org/pil/19.2.html
BTW, do you have a more precise and more formal/complete doc about lua than this website ?

Thanks for changing the code. As soon as it will be released, I will erase my "patch".

**Comment by jtackaberry** [comment]
>  BTW, do you have a more precise and more formal/complete doc about lua than this website ?

I wish I did.  I've never been terribly impressed by Lua's documentation and I don't think I've ever seen a formal specification.

So it's off to Lua's source code, and the relevant function is:

https://github.com/lua/lua/blob/v5.3.6/ltablib.c#L79:L103

There we can see that `e` is set to, effectively, `#t+1` and with the 2-argument append-to-table form of `table.insert()`, `e` is used, but you can also explicitly pass this index in the 3-argument form.  The bounds checking allows the index to be `e` but not greater.

In other words, these two lines are semantically equivalent:

```lua
t[#t+1] = 'foo'
table.insert(t, #t+1, 'foo')
```

> Thanks for changing the code. As soon as it will be released, I will erase my "patch".

It's out now :)

**Comment by DaveInDev** [comment]
Just tested and it works fine, thanks. Again our posting crossed at the same moment in the abysses of the Atlantic Ocean.
